pkg/controller/daemon/daemoncontroller.go

```diff
@@ -899,11 +924,16 @@ func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *v1.Node, ds *exten
 		return false, false, false, err
 	}
 
+	// Check if need to fire InsufficientResourceError
+	fireInsufficientResourceError := shouldFireInsufficientResourceError(reasons)
+
 	for _, r := range reasons {
 		glog.V(4).Infof("DaemonSet Predicates failed on node %s for ds '%s/%s' for reason: %v", node.Name, ds.ObjectMeta.Namespace, ds.ObjectMeta.Name, r.GetReason())
 		switch reason := r.(type) {
 		case *predicates.InsufficientResourceError:
-			dsc.eventRecorder.Eventf(ds, v1.EventTypeWarning, FailedPlacementReason, "failed to place pod on %q: %s", node.ObjectMeta.Name, reason.Error())
+			if fireInsufficientResourceError {
+				dsc.eventRecorder.Eventf(ds, v1.EventTypeWarning, FailedPlacementReason, "failed to place pod on %q: %s", node.ObjectMeta.Name, reason.Error())
+			}
 			shouldSchedule = false
```

Comment:
>I'm not sure if I understand this part. If the Predicates returns both InsufficientResourceError and ErrNodeSelectorNotMatch, we'll just hit the ErrNodeSelectorNotMatch and if we return immediately we're done. We won't emit "failed to place pod" event, isn't this what we want?

We can take a use case as in https://github.com/kubernetes/kubernetes/issues/45780
1) I have two nodes one master and one worker, master node has 10 cpu and 10 memory, worker node has 100 cpu and 100 memory with a label as `foo=bar`.
2) Then I have a demonset with `nodeSelector` as `foo: bar` and also request resources as 20 cpu and 20 memory.
3) After deployment, one pod for the daemonset will be running in worker node, this is what I expected.
4) The problem is that the daemonset event will report it `failed to place pod on "<node>": Node didn't have enough resource: memory, requested: 9768534016, used: 8361345024, capacity: 14944219136` as the master node do not have enough resources, but this is not I want as I have specified the `nodeSelector` for the daemonset.

The fix here is trying to avoid some confusing event, hope this helps ;-)

>Note that we didn't return immediately here, but continue the loop. If we just want to skip emitting "failed to place pod" event when there's an intentional error, why not just return immediately?

You can probably take a look at https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L883-L902 , for master node that I listed above, we will check `noncriticalPredicates` first and then `EssentialPredicates`, the `noncriticalPredicates` will return `InsufficientResourceError` and `EssentialPredicates` will return `ErrNodeSelectorNotMatch`. In the failure reason queue, `InsufficientResourceError` is ahead of `ErrNodeSelectorNotMatch`, so in the switch case here, we will always fire the `InsufficientResourceError` event first. It may not help even if we return directly for `intentional` reasons as this case will be handled after `InsufficientResourceError`.
I want to switch the order of `noncriticalPredicates` and `EssentialPredicates` in https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L883-L902, but it may not be good as we do not want to update the scheduler logic due to daemonset requirement, but should update daemonset to handle such case.